### PR TITLE
Generate numbers image with CLI

### LINKED ISSUE #512: Brainstorm about an improved way to generate numbers
Following minutes of meeting with Alan, I found that we need to find a better way to generate the numbers in FSLeyes.

I am quoting the notes taken from the meeting below,

```
- have a way to generate them outside of FSLeyes
- increase the size of the numbers, or find another way to generate them. Eg: output image with/without the numbers. 
```

It would be nice if some context can be provided about the issue and we can start discussing this, hence I opened this issue to keep the discussion get started.

### PR DESCRIPTION
**Pull request by Stoyan-I-A**
## Checklist

- [X] I've given this PR a concise, self-descriptive, and meaningful title
- [X] I've linked relevant issues in the PR body
- [X] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've added relevant tests for my contribution
- [X] I've updated the documentation and/or added correct docstrings
- [x] I've assigned a reviewer
- [x] I've consulted [ADS's internal developer documentation](https://github.com/neuropoly/axondeepseg/wiki/Guidelines-for-contribution) to ensure my contribution is in line with any relevant design decisions

## Description
- This PR implements a change in the CLI `axondeepseg_morphometrics`
- This change generates two output files with the following suffixes: 
  - `_index.png`: Only numbers
  - `_axonmyelin_index.png`: Axon+myelin seg with numbers on it.
- In the FSLeyes plugin, the index and axonmyelin+index images will be saved in the same repository as the output Excel file

## Linked issues
Resolves #512

### CODE REVIEW

**Review by vs74 — COMMENTED** [review]
As described, this PR works amazingly in generating numbers named `numbers.png` with both CLI and FSLeyes. 

I tested this feature on **test/__test_files__/__test_demo_files__/image.png**

## **image.png**
image

## **numbers.png**

image


I have two suggestions to make though

1. First, instead of having `numbers.png` image, I guess it would be better to rename it something like `image_name_numbers.png` because when we compute the morphometrics in batches of the images present in the same directory, there would be only one `numbers.png` file generated as it would be overwritten. 

2. Second, instead of only just generating numbers of an image, it would be a better idea to generate overlaid numbers on the axon and myelin objects because when there are a lot of such axon myelin objects it becomes increasingly difficult to just look at the `numbers.png` and decided which instance belongs to which instance in the  axon myelin mask. 

     Also, users, would less likely overlay the generated `numbers.png` on the segmented 
     axonmyelin mask and they would expect us to do that. 

So, I believe an overlaid axonmyelin mask on top of numbers would be better. 

Perhaps something like this : 
## numbers_seg.png
image

**Comment by vs74** [comment]
Ignore my second suggestion, we would be requiring overlaid numbers on the EM image instead of just the numbers, especially for Alan's project. Is that possible?

**Comment by mariehbourget** [comment]
> Ignore my second suggestion, we would be requiring overlaid numbers on the EM image instead of just the numbers, especially for Alan's project. Is that possible?

This may be tricky because of the color of the numbers vs. the color of the image background, depending on the modality, the background can be more black or more white. We could implement it for Alan's project if he needs it but I don't think that generating an original image with overlay numbers on it should be a permanent feature. Also, converting the original image from grayscale to color to overlay the numbers will generate a lot of data. That could be partially alleviated by saving the new files as jpeg as they are for QC only.

I'm in favor or your suggestion 2 with "numbers_seg.png" as a permanent feature (as we did last week for Alan's project). As you mentioned: "users would less likely overlay the generated `numbers.png` on the segmented axonmyelin mask and they would expect us to do that". @vs74, can you share with Stoyan the script you used to generate that in the CLI last week?

**Comment by vs74** [comment]
> This may be tricky because of the color of the numbers vs. the color of the image background, depending on the modality, the background can be more black or more white. We could implement it for Alan's project if he needs it but I don't think that generating an original image with overlay numbers on it should be a permanent feature. Also, converting the original image from grayscale to color to overlay the numbers will generate a lot of data. That could be partially alleviated by saving the new files as jpeg as they are for QC only.

I agree with you @mariehbourget. Perhaps we have to modify our scripts if Alan asks us to overlay EM images on top of numbers. 

> I'm in favor or your suggestion 2 with "numbers_seg.png" as a permanent feature (as we did last week for Alan's project). As you mentioned: "users would less likely overlay the generated numbers.png on the segmented axonmyelin mask and they would expect us to do that". @vs74, can you share with Stoyan the script you used to generate that in the CLI last week?

Here is the modified **launch_morphometrics_computation.py**  script which me and @mariehbourget created for generating `_numbers_seg.png` numbers overlaid on axonmyelin masks. This script also takes care of my suggestion 1, which is renaming `numbers.png` by adding `_numbers_seg.png`  as a suffix to the image name:

<summary>
<details>

```
# coding: utf-8

from pathlib import Path
import argparse
from argparse import RawTextHelpFormatter
from matplotlib import image
import sys

# Scientific modules imports
import numpy as np
import pandas as pd

# AxonDeepSeg imports
from AxonDeepSeg.morphometrics.compute_morphometrics import (
                                                                get_axon_morphometrics, 
                                                                save_axon_morphometrics,  
                                                                draw_axon_diameter,
                                                                save_map_of_axon_diameters,
                                                                get_aggregate_morphometrics,
                                                                write_aggregate_morphometrics 
                                                            )
import AxonDeepSeg.ads_utils as ads
from config import axon_suffix, myelin_suffix
from AxonDeepSeg.ads_utils import convert_path

from PIL import Image, ImageOps
from AxonDeepSeg.postprocessing import generate_axon_numbers_image
from config import axonmyelin_suffix

def launch_morphometrics_computation(path_img, path_prediction):
    """
    This function is equivalent to the morphometrics_extraction notebook of AxonDeepSeg.
    It automatically performs all steps (computations, savings, displays,...) of the
    morphometrics extraction of a given sample.
    :param path_img: path of the input image (microscopy sample)
    :param path_prediction: path of the segmented image (output of AxonDeepSeg)
    :return: none.
    """
    
    # If string, convert to Path objects
    path_img = convert_path(path_img)
    path_prediction = convert_path(path_prediction)

    try:
        # Read image
        img = ads.imread(path_img)

        # Read prediction
        pred = ads.imread(path_prediction)
    except (IOError, OSError) as e:
        print(("launch_morphometrics_computation: " + str(e)))
        raise
    else:

        # Get axon and myelin masks
        pred_axon = pred > 200
        pred_myelin = np.logical_and(pred >= 50, pred <= 200)

        # Get folder path
        path_folder = path_img.parent

        # Compute and save axon morphometrics
        stats_array = get_axon_morphometrics(pred_axon, path_folder)
        save_axon_morphometrics(path_folder, stats_array)

        # Generate and save displays of axon morphometrics
        fig = draw_axon_diameter(img, path_prediction, pred_axon, pred_myelin)
        save_map_of_axon_diameters(path_folder, fig)

        # Compute and save aggregate morphometrics
        aggregate_metrics = get_aggregate_morphometrics(
            pred_axon, pred_myelin, path_folder
        )
        write_aggregate_morphometrics(path_folder, aggregate_metrics)


def main(argv=None):
    ap = argparse.ArgumentParser(formatter_class=RawTextHelpFormatter)

    # Setting the arguments of the saving the morphometrics in excel file
    ap.add_argument('-s', '--sizepixel', required=False, help='Pixel size of the image(s) to compute morphometrics, in micrometers. \n' +
                                                              'If no pixel size is specified, a pixel_size_in_micrometer.txt \n' +
                                                              'file needs to be added to the image folder path. The pixel size \n' +
                                                              'in that file will be used for the morphometrics computation.',
                                                              default=None)

    ap.add_argument('-i', '--imgpath', required=True, nargs='+', help='Path to the image.')

    ap.add_argument('-f', '--filename', required=False,  help='Name of the excel file in which the morphometrics will be stored',
                                                              default="axon_morphometrics")

    # Processing the arguments
    args = vars(ap.parse_args(argv))
    path_target_list = [Path(p) for p in args["imgpath"]]
    filename = str(args["filename"])

    # Tuple of valid file extensions
    validExtensions = (
                        ".jpeg",
                        ".jpg",
                        ".tif",
                        ".tiff",
                        ".png"
                        )

    for current_path_target in path_target_list:
        if current_path_target.suffix.lower() in validExtensions:
            # load the axon mask
            if (Path(str(current_path_target.with_suffix("")) + str(axon_suffix))).exists():
                pred_axon = image.imread(str(current_path_target.with_suffix("")) + str(axon_suffix))
            else: 
                print("ERROR: Segmented axon mask is not present in the image folder. ",
                              "Please check that the axon mask is located in the image folder. ",
                              "If it is not present, perform segmentation of the image first using ADS."
                      )
                sys.exit(3)

            # load myelin mask    
            if (Path(str(current_path_target.with_suffix("")) + str(myelin_suffix))).exists():
                pred_myelin = image.imread(str(current_path_target.with_suffix("")) + str(myelin_suffix))
            else: 
                print("ERROR: Segmented myelin mask is not present in the image folder. ",
                              "Please check that the myelin mask is located in the image folder. ",
                              "If it is not present, perform segmentation of the image first using ADS."
                      )
                sys.exit(3)

            if args["sizepixel"] is not None:
                psm = float(args["sizepixel"])
            else:  # Handle cases if no resolution is provided on the CLI

                # Check if a pixel size file exists, if so read it.
                if (current_path_target.parent / 'pixel_size_in_micrometer.txt').exists():

                    resolution_file = open(current_path_target.parent / 'pixel_size_in_micrometer.txt', 'r')

                    psm = float(resolution_file.read())
                else:

                    print("ERROR: No pixel size is provided, and there is no pixel_size_in_micrometer.txt file in image folder. ",
                                  "Please provide a pixel size (using argument -s), or add a pixel_size_in_micrometer.txt file ",
                                  "containing the pixel size value."
                          )
                    sys.exit(3)

            x = np.array([], dtype=[
                                        ('x0', 'f4'),
                                        ('y0', 'f4'),
                                        ('gratio', 'f4'),
                                        ('axon_area', 'f4'),
                                        ('axon_perimeter', 'f4'),
                                        ('myelin_area', 'f4'),
                                        ('axon_diam', 'f4'),
                                        ('myelin_thickness', 'f4'),
                                        ('axonmyelin_area', 'f4'),
                                        ('axonmyelin_perimeter', 'f4'),
                                        ('solidity', 'f4'),
                                        ('eccentricity', 'f4'),
                                        ('orientation', 'f4')
                                    ]
                         )
            
            # Compute statistics
            stats_array = get_axon_morphometrics(im_axon=pred_axon, im_myelin=pred_myelin, pixel_size=psm)

            for stats in stats_array:

                x = np.append(x, np.array(
                        [(
                            stats['x0'],
                            stats['y0'],
                            stats['gratio'],
                            stats['axon_area'],
                            stats['axon_perimeter'],
                            stats['myelin_area'],
                            stats['axon_diam'],
                            stats['myelin_thickness'],
                            stats['axonmyelin_area'],
                            stats['axonmyelin_perimeter'],
                            stats['solidity'],
                            stats['eccentricity'],
                            stats['orientation']
                        )],
                        dtype=x.dtype)
                    )

            # Generate numbers layers and png file with overlay

            mean_diameter_in_pixel = np.average(x['axon_diam']) / psm
            axon_indexes = np.arange(x.size)
            numbers = generate_axon_numbers_image(axon_indexes, x['x0'], x['y0'],
                                                                  tuple(reversed(pred_axon.shape)),
                                                                  mean_diameter_in_pixel)
            seg = Image.open(str(current_path_target.with_suffix("")) + str(axonmyelin_suffix))
            overlay = Image.fromarray(numbers)           
            numbers_file = ImageOps.colorize(seg, black="black", white="blue", mid="red", blackpoint=0, whitepoint=255, midpoint=127)
            numbers_file.paste(overlay, (0, 0), overlay)
            numbers_file.save(current_path_target.parent / (str(current_path_target.with_suffix("")) + "_numbers_seg.png"),"PNG")

            # save the current contents in the file
            if not (filename.lower().endswith((".xlsx", ".csv"))):  # If the user didn't add the extension, add it here
                filename = filename + '.xlsx' 
            try:
                # Export to excel
                if filename.endswith('.xlsx'):
                    pd.DataFrame(x).to_excel(current_path_target.parent / filename)
                # Export to csv    
                else: 
                    pd.DataFrame(x).to_csv(current_path_target.parent / filename)
                    
                print(f"Moprhometrics file: {filename} has been saved in the {str(current_path_target.parent.absolute())} directory")
            except IOError:
                print("Cannot save morphometrics data in file '%s'." % filename)

        else: 
            print("The path(s) specified is/are not image(s). Please update the input path(s) and try again.")
            break    

    sys.exit(0)
# Calling the script
if __name__ == '__main__':
    main()
```
</details>
</summary>

Let me know if you have any questions in regards to the script.

**Comment by Stoyan-I-A** [comment]
All right, I should be able to update this PR before Monday.

**Comment by Stoyan-I-A** [comment]
Thanks for the script, I pretty much copied the logic. You can check this PR again now.

File: AxonDeepSeg/morphometrics/launch_morphometrics_computation.py
Comment (vs74) [inline AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:200]: Would it be a better idea to add the suffix `_numbers_seg.png` in the [config](https://github.com/neuropoly/axondeepseg/blob/master/config.py) file rather than hardcoding?

File: AxonDeepSeg/morphometrics/launch_morphometrics_computation.py
Comment (vs74) [inline AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:205]: Same for the suffix `_colored_seg.png`, should we add it in the [config](https://github.com/neuropoly/axondeepseg/blob/master/config.py) file?

File: ads_plugin.py
Comment (vs74) [inline ads_plugin.py:647]: I wasn't able to test this PR for FSLeyes because of issue #521.  

Looking by the code, I'm not sure whether the FSLeyes also generates two files: `_numbers_seg.png` and `_colored_seg.png`? The behaviour should be consistent with FSLeyes and CLI.

**Review by vs74 — COMMENTED** [review]
I tested this PR for the CLI and it works as expected. It generates two numbered files: `_numbers_seg.png` (numbers only) and `_colored_seg.png` (numbers + overlaid segs).

I'm not sure we want to go with saving `_numbers_seg.png` (numbers only) as pointed by @mariehbourget  earlier. So we need to discuss about this in our meeting. 

@Stoyan-I-A could you add some tests for CLI?

For FSLeyes, because of issue #521, I wasn't able to test the functionality. Perhaps could someone test this PR for FSLeyes?

File: AxonDeepSeg/morphometrics/launch_morphometrics_computation.py
Comment (jcohenadad) [inline AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:210]: Moprhometrics --> Morphometrics

**Review by jcohenadad — CHANGES_REQUESTED** [review]
Suggest to change name for:
- `IMAGE_index.png`
- `AXONMYELIN_index.png`

We will also need to update the output name in FSLeyes plugin.

**Comment by vs74** [comment]
In b4245e9, I've addressed the merge conflicts after merging #399.

**Comment by Stoyan-I-A** [comment]
The FSLeyes output and the naming of the files have been addressed. 
However, this PR is still not ready for review.
TODO:
- Write tests
- Add docstring
- Find out why current tests are not passing.

**Comment by Stoyan-I-A** [comment]
This PR is ready for review

**Comment by jcohenadad** [comment]
@Stoyan-I-A sorry my comment was misleading, i suggest to not use capital letters as suffix

File: AxonDeepSeg/postprocessing.py
Comment (vasudev-sharma) [inline AxonDeepSeg/postprocessing.py:139]: @Stoyan-I-A, I believe the tests for this new function (`generate_and_save_colored_image_with_index_numbers`) are missing from the tests. Could you write unit tests for this as well so that we can be sure that it works as expected?

File: config.py
Comment (vasudev-sharma) [inline config.py:8]: As discussed in the recent meeting, we might want to change the convention of the suffix file names (`index_suffix` and `axonmyelin_index_suffix`) from CAPITAL letters to small letters. Just adding this comment out here, so I don't miss this while reviewing this PR.

File: ads_plugin.py
Comment (vasudev-sharma) [inline ads_plugin.py:36]: This is not specifically related to this PR, but recently I merged https://github.com/neuropoly/axondeepseg/pull/487 to avoid using wild card imports in the scripts. I guess I might have skipped this script and the specific import (
`from AxonDeepSeg.morphometrics.compute_morphometrics import *`) while we merged my PR. 

Would it be possible to remove the wild card import in this script so that this bad practise is not followed and the codebase is consistent?

Thanks!

**Comment by vasudev-sharma** [comment]
Functionality-wise, both the FLSeyes and CLI are working as expected, that is, the `indices` and `indices_seg` are saved after computing the morphometrics in the directory where `morphometrics.csv` file is saved. Thanks @Stoyan-I-A  for this contribution. This PR is good to go after review comments on this PR have been resolved. 

One question:  In one of our meetings, we briefly discussed about having `indices` image saved or not, but I vaguely remember what did we discuss about it. In particular, did we decide to go ahead with saving `indices` image as well by default?

File: AxonDeepSeg/postprocessing.py
Comment (Stoyan-I-A) [reply AxonDeepSeg/postprocessing.py:139]: You're right, I will write a test for this function. It might require adding some new files to the `test_data `however.

File: ads_plugin.py
Comment (Stoyan-I-A) [reply ads_plugin.py:36]: Yeah I agree we should remove it.

**Comment by Stoyan-I-A** [comment]
> In one of our meetings, we briefly discussed about having `indices` image saved or not, but I vaguely remember what did we discuss about it. In particular, did we decide to go ahead with saving `indices` image as well by default?

As far as I remember, we were still not sure if want to keep both images or just the colored one, so I decided to keep both for the moment.

**Comment by Stoyan-I-A** [comment]
I need to add two new images to the `__test_files__`. I can provide them right now or I can wait for #533 to be merged and add them myself.

**Comment by vasudev-sharma** [comment]
I believe #533 is only related to models and https://github.com/axondeepseg/data-testing would be where `__test_ files__` where these extra 2 images need to added. @hermancollin correct me if I'm wrong : the test files link in the ADS repo hasn't been changed right in the PR #533?

**Comment by hermancollin** [comment]
> I believe #533 is only related to models and https://github.com/axondeepseg/data-testing would be where `__test_ files__` are present, where these extra 2 images needs to added. @hermancollin correct me if I'm wrong : the test files link in the ADS repo hasn't been changed in the PR #533?

That is correct, this PR is for models only. @Stoyan-I-A if you want to add files to the testing folder, you can directly commit to the master branch here: https://github.com/axondeepseg/data-testing (or I guess the "correct" way would be to start a branch, then do a PR).

**Comment by vasudev-sharma** [comment]
Also, we need to open another PR to update the link of `__test_files__` in ADS and link it to upload test files on  https://github.com/axondeepseg/data-testing

**Comment by vasudev-sharma** [comment]
> Also, we need to open another PR to update the link of `__test_files__` in ADS and link it to uploaded test files on https://github.com/axondeepseg/data-testing

@Stoyan-I-A I've created PR #534 for the same. If you wish to upload the 2 additional images, you can create a PR in https://github.com/axondeepseg/data-testing repo or if you want me to upload them, I can do the same.

File: config.py
Comment (vasudev-sharma) [inline config.py:8]: I would perhaps name it `_index.png`. The reason is we can have an image named `image.png`. And after the morphometrics computation, the index file would be named as `image_image_index.png` which doesn't looks very intuitive.

**Review by vasudev-sharma — COMMENTED** [review]
Looks like there is inconsistency in the naming of `index` and `axonmyelin_index` file for FSLeyes and CLI.
---
**CLI**
image
---
**FSLeyes**
image

File: config.py
Comment (Stoyan-I-A) [reply config.py]: Done!

**Review by vasudev-sharma — APPROVED** [review]
Thanks @Stoyan-I-A for your contribution 🥳. Everything is working as expected

**Comment by Stoyan-I-A** [comment]
@jcohenadad I requested a review because GitHub doesn't let me merge since you have made a change request a few weeks ago. Those change have been made however I can't mark them as such.

**Review by jcohenadad — APPROVED** [review]
Awesome contribution, thank you @Stoyan-I-A 👍

**Pull request merged**
